File: flutter/lib/models/published_ontology.dart
// 공표된 온톨로지 관련 모델 (CIDOC-CRM, Schema.org, FOAF, Dublin Core 등)

class ClassMapping {
  final String curie;       // e.g. "crm:E21_Person"
  final String ontologyId;  // e.g. "cidoc-crm"
  final bool isPrimary;
  final double confidence;
  final String note;

  const ClassMapping({
    required this.curie,
    required this.ontologyId,
    this.isPrimary = true,
    this.confidence = 1.0,
    this.note = '',
  });

  factory ClassMapping.fromJson(Map<String, dynamic> json) => ClassMapping(
        curie: json['curie'] as String? ?? '',
        ontologyId: json['ontology_id'] as String? ?? '',
        isPrimary: json['is_primary'] as bool? ?? true,
        confidence: (json['confidence'] as num?)?.toDouble() ?? 1.0,
        note: json['note'] as String? ?? '',
      );

  Map<String, dynamic> toJson() => {
        'curie': curie,
        'ontology_id': ontologyId,
        'is_primary': isPrimary,
        'confidence': confidence,
        'note': note,
      };

  ClassMapping copyWith({
    String? curie,
    String? ontologyId,
    bool? isPrimary,
    double? confidence,
    String? note,
  }) =>
      ClassMapping(
        curie: curie ?? this.curie,
        ontologyId: ontologyId ?? this.ontologyId,
        isPrimary: isPrimary ?? this.isPrimary,
        confidence: confidence ?? this.confidence,
        note: note ?? this.note,
      );
}

class PublishedClass {
  final String id;
  final String curie;
  final String uri;
  final String label;
  final String labelKo;
  final String description;

  const PublishedClass({
    required this.id,
    required this.curie,
    required this.uri,
    required this.label,
    required this.labelKo,
    this.description = '',
  });

  factory PublishedClass.fromJson(Map<String, dynamic> json) => PublishedClass(
        id: json['id'] as String? ?? '',
        curie: json['curie'] as String? ?? '',
        uri: json['uri'] as String? ?? '',
        label: json['label'] as String? ?? '',
        labelKo: json['label_ko'] as String? ?? '',
        description: json['description'] as String? ?? '',
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'curie': curie,
        'uri': uri,
        'label': label,
        'label_ko': labelKo,
        'description': description,
      };

  PublishedClass copyWith({
    String? curie,
    String? uri,
    String? label,
    String? labelKo,
    String? description,
  }) =>
      PublishedClass(
        id: id,
        curie: curie ?? this.curie,
        uri: uri ?? this.uri,
        label: label ?? this.label,
        labelKo: labelKo ?? this.labelKo,
        description: description ?? this.description,
      );
}

class PublishedOntology {
  final String id;
  final String name;
  final String prefix;
  final String namespaceUri;
  final String description;
  final String version;
  final List<PublishedClass> classes;

  const PublishedOntology({
    required this.id,
    required this.name,
    required this.prefix,
    required this.namespaceUri,
    this.description = '',
    this.version = '',
    this.classes = const [],
  });

  factory PublishedOntology.fromJson(Map<String, dynamic> json) =>
      PublishedOntology(
        id: json['id'] as String? ?? '',
        name: json['name'] as String? ?? '',
        prefix: json['prefix'] as String? ?? '',
        namespaceUri: json['namespace_uri'] as String? ?? '',
        description: json['description'] as String? ?? '',
        version: json['version'] as String? ?? '',
        classes: (json['classes'] as List? ?? [])
            .map((c) => PublishedClass.fromJson(c as Map<String, dynamic>))
            .toList(),
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'prefix': prefix,
        'namespace_uri': namespaceUri,
        'description': description,
        'version': version,
        'classes': classes.map((c) => c.toJson()).toList(),
      };
}

File: ontology/published_ontology.py
"""ontology/published_ontology.py — 공표된 온톨로지 데이터 모델"""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _short_id() -> str:
    return uuid.uuid4().hex[:12]


@dataclass
class PublishedClass:
    """공표된 온톨로지의 개별 클래스."""

    curie: str       # e.g. "crm:E21_Person"
    uri: str         # e.g. "http://www.cidoc-crm.org/cidoc-crm/E21_Person"
    label: str       # English label, e.g. "Person"
    label_ko: str    # 한국어 레이블, e.g. "인물"
    description: str = ""
    id: str = field(default_factory=_short_id)

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "curie": self.curie,
            "uri": self.uri,
            "label": self.label,
            "label_ko": self.label_ko,
            "description": self.description,
        }

    @staticmethod
    def from_dict(d: dict) -> "PublishedClass":
        return PublishedClass(
            id=d.get("id", _short_id()),
            curie=d["curie"],
            uri=d["uri"],
            label=d["label"],
            label_ko=d.get("label_ko", ""),
            description=d.get("description", ""),
        )


@dataclass
class PublishedOntology:
    """공표된 온톨로지 (CIDOC-CRM, Schema.org, FOAF, Dublin Core 등)."""

    id: str            # slug, e.g. "cidoc-crm"
    name: str          # full name, e.g. "CIDOC Conceptual Reference Model"
    prefix: str        # namespace prefix, e.g. "crm"
    namespace_uri: str # base URI, e.g. "http://www.cidoc-crm.org/cidoc-crm/"
    description: str = ""
    version: str = ""
    classes: list[PublishedClass] = field(default_factory=list)
    created_at: str = field(default_factory=_now)
    updated_at: str = field(default_factory=_now)

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "prefix": self.prefix,
            "namespace_uri": self.namespace_uri,
            "description": self.description,
            "version": self.version,
            "classes": [c.to_dict() for c in self.classes],
            "created_at": self.created_at,
            "updated_at": self.updated_at,
        }

    @staticmethod
    def from_dict(d: dict) -> "PublishedOntology":
        return PublishedOntology(
            id=d["id"],
            name=d["name"],
            prefix=d["prefix"],
            namespace_uri=d["namespace_uri"],
            description=d.get("description", ""),
            version=d.get("version", ""),
            classes=[PublishedClass.from_dict(c) for c in d.get("classes", [])],
            created_at=d.get("created_at", _now()),
            updated_at=d.get("updated_at", _now()),
        )
